crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3276,6 +3294,23 @@ impl<'db> TypeInferenceBuilder<'db> {
                     value_ty,
                     Type::IntLiteral(i64::from(bool)),
                 ),
+
+            // Handling of Special Forms
+            (Type::ClassLiteral(class), slice_ty)
+                if class.is_known(self.db, KnownClass::SpecialForm) =>
+            {
+                if class.name(self.db) == "Literal" {
+                    match slice_ty {
+                        Type::Tuple(tuple) => {
+                            let elts = tuple.elements(self.db);
+                            Type::Union(UnionType::new(self.db, elts))
+                        }
+                        ty => ty,
+                    }
+                } else {
+                    Type::Todo
+                }
+            }
```

Comment:
I think here we should not be handling special forms at all, since this is type inference for value expressions, not type expressions. (This is related to the comment on the tests above.)